### PR TITLE
C#: Implement CFG for `not` patterns

### PR DESCRIPTION
**Pull request by hvitved**
This PR changes the CFG for pattern expressions (including `not` expressions). Example:

```csharp
public static bool M3(object c) =>
    c is not null ? c is 1 : c is 2;
```

<details>
  <summary>CFG before:</summary>
  
  before
</details>

<details>
  <summary>CFG after:</summary>
  
  after
</details>

~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/782/~
~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/799/~
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/804/

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll [hunk 1]
```diff
@@ -232,37 +232,61 @@ private Expr getQualifier(QualifiableExpr e) {
...
-        e = tpe and
-        t = tpe.getCheckedType() and
-        strippedType = stripped.getType() and
-        not t.isImplicitlyConvertibleTo(strippedType) and
-        not t instanceof Interface and
-        not t.containsTypeParameters() and
-        not strippedType.containsTypeParameters() and
-        value = false
+private predicate isMatchingConstant(PatternExpr pe, boolean value) {
+  exists(Expr e | mustHaveMatchingCompletion(e, pe) |
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll:272]: Is there any benefit of having nested `exists` vs adding all variables into a single one?

**Review by tamasvajk — APPROVED** [review]
LGTM

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll (hunk 1 above)
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll:272]: No. I'll merge them into one.

**Review by tamasvajk — APPROVED** [review]

**Comment by hvitved** [comment]
Rebased and added two new commits.

**Review by tamasvajk — APPROVED** [review]

**Comment by hvitved** [comment]
Fixed a join-order regression caused by 355edcb13699187cf611061c803848a2494ae083, and started a new differences job.

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/ControlFlowGraphImpl.qll [hunk 2]
```diff
@@ -1372,11 +1377,16 @@ module Statements {
...
+      exists(int nestLevel |
+        c =
+          any(NestedCompletion nc |
+            this.lastFinally(last, nc.getAnInnerCompatibleCompletion(), nc.getOuterCompletion(),
+              nestLevel) and
+            // unbind
+            nc.getNestLevel() >= nestLevel and
+            nc.getNestLevel() <= nestLevel
+          )
+      )
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/controlflow/internal/ControlFlowGraphImpl.qll:1389]: This seems overly complicated. Can you show me what perf issue was solved by this?

**Review by tamasvajk — APPROVED** [review]

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/ControlFlowGraphImpl.qll (hunk 2 above)
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/internal/ControlFlowGraphImpl.qll:1389]: You are right, it is complicated. It solves a bad-join order that was also there before this PR, but which I spotted when fixing the other bad join-order:
```
                      1986     ~0%       {5} r84 = SCAN ControlFlowGraphImpl::Statements::TryStmtTree::lastFinally_dispred#fffff#prev_delta AS I OUTPUT I.<4>, I.<0> 'this', I.<1> 'last', I.<2> 'c', I.<3>
                      29295324 ~1%       {5} r85 = JOIN r84 WITH Completion::TNestedCompletion#ffff_23#join_rhs AS R ON FIRST 1 OUTPUT R.<1> 'c', r84.<4>, r84.<1> 'this', r84.<2> 'last', r84.<3>
                      10451    ~0%       {4} r86 = JOIN r85 WITH Completion::NestedCompletion::getOuterCompletion_dispred#ff AS R ON FIRST 2 OUTPUT r85.<0> 'c', r85.<4>, r85.<2> 'this', r85.<3> 'last'
                      499      ~3%       {3} r87 = JOIN r86 WITH Completion::NestedCompletion::getAnInnerCompatibleCompletion_dispred#ff AS R ON FIRST 2 OUTPUT r86.<3> 'last', r86.<0> 'c', r86.<2> 'this'
```
The rewrite avoids joining first on `getNestLevel()`, so it instead becomes
```
                      1986    ~0%       {5} r93 = SCAN ControlFlowGraphImpl::Statements::TryStmtTree::lastFinally_dispred#fffff#prev_delta AS I OUTPUT I.<3>, I.<0> 'this', I.<1> 'last', I.<2> 'c', I.<4>
                      28433   ~0%       {5} r94 = JOIN r93 WITH Completion::NestedCompletion::getOuterCompletion_dispred#ff_10#join_rhs AS R ON FIRST 1 OUTPUT R.<1> 'c', r93.<3>, r93.<1> 'this', r93.<2> 'last', r93.<4>
                      1497    ~7%       {4} r95 = JOIN r94 WITH Completion::NestedCompletion::getAnInnerCompatibleCompletion_dispred#ff AS R ON FIRST 2 OUTPUT r94.<0> 'c', r94.<2> 'this', r94.<3> 'last', r94.<4>
                      1497    ~0%       {7} r96 = JOIN r95 WITH Completion::TNestedCompletion#ffff_3012#join_rhs AS R ON FIRST 1 OUTPUT r95.<1> 'this', r95.<2> 'last', r95.<3>, r95.<0> 'c', R.<1>, R.<2>, R.<3>
                      1497    ~0%       {7} r97 = SELECT r96 ON r96.<6> >= r96.<2>
                      499     ~0%       {7} r98 = SELECT r97 ON r97.<6> <= r97.<2>
                      499     ~3%       {3} r99 = SCAN r98 OUTPUT r98.<1> 'last', r98.<3> 'c', r98.<0> 'this'
```

**Pull request merged**